### PR TITLE
stability handler

### LINKED ISSUE #7729: [Stability AI] Integrate Stability AI with mindsdb
### Short description and motivation for the proposed feature

The goal of this issue is to build a new handler that will connect MindsDB to Stability AI.

### Resources 💡

🏁 To get started, please refer to the following resources:

📚 [Building Data Handler docs](https://docs.mindsdb.com/contribute/data-handlers): This documentation will guide you through the process of building a custom handler for MindsDB. It contains essential information on the handler structure, methods, and best practices.

📖 [Stability AI](https://stability.ai/)

👉 [Monkey Learn](https://github.com/mindsdb/mindsdb/blob/staging/mindsdb/integrations/handlers/monkeylearn_handler/README.md): This existing handler connects MindsDB with MonkeyLearn.

### Next Steps 🕵️‍♂️ 🕵️‍♀️

You can familiarize yourself with Stability AI.
Create a new handler that connects MindsDB to Stability AI.
Implement the core features mentioned above.
Thoroughly test the integration to ensure proper functionality.
Submit a pull request with the new Stability AI Integration handler.

### 📢 Additional rewards 🏆

This integration brings 7️⃣ 0️⃣ : points for SWAG and entry into the draw for a 👇 :

💻 [Razer Blade 16 Laptop](https://www.razer.com/gaming-laptops/Razer-Blade-16/RZ09-0483SEJ3-R3U1)
👕 🐻 [MindsDB Swag](https://store.covver.io/mindsdb/collections/swag-collection)
ℹ️ For more info check out https://mindsdb.com/hacktoberfest/ 👈

### PR DESCRIPTION
**Pull request by Better-Boy**
## Description

Please include a summary of the change and the issue it solves. 

Fixes #7729 

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [ ] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Review by lnxpy — CHANGES_REQUESTED** [review]
I'm not sure if a `setup.py` file would come in handy for a handler at this point. All you need to have is a `requirements.txt` that contains your dependencies. This file might cause some misunderstandings. Some `setup()` parameters are valued wrongly such as `name`.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+from stability_sdk import client
+from stability_sdk.client import generation
+from PIL import Image
+import io, requests
+
+class StabilityAPIClient:
+    
+    def __init__(self, api_key, dir_to_save, 
+                 engine="stable-diffusion-xl-1024-v1-0", upscale_engine="esrgan-v1-x2plus"):
+        self.api_key = api_key
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:10]: Please mention in the comments about what this function does.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            raise ValueError("Unknown engine. The available engines are - " + self.available_engines)
+        if not self._is_valid_engine(upscale_engine):
+            raise ValueError("Unknown upscale engine. The available engines are - " + self.available_engines)
+        
+        self.stability_api = client.StabilityInference(host=self.STABILITY_HOST, 
+                                                       key=self.api_key,
+                                                       engine=engine,
+                                                       upscale_engine=upscale_engine,
+                                                       verbose=True)
+
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:24]: Please wrap this client request in a try-except block if error, then raise and exception.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            raise ValueError("Unknown upscale engine. The available engines are - " + self.available_engines)
+        
+        self.stability_api = client.StabilityInference(host=self.STABILITY_HOST, 
+                                                       key=self.api_key,
+                                                       engine=engine,
+                                                       upscale_engine=upscale_engine,
+                                                       verbose=True)
+        
+    def save_image(self, artifact):
+        img = Image.open(io.BytesIO(artifact.binary))
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:26]: Please mention in the comments about what this function does.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        path = self.save_dir + str(artifact.seed)+ ".png"
+        img.save(path)
+        return path
+    
+    def _process_artifacts(self, artifacts):
+        saved_image_paths = []
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:32]: Please mention in the comments about what this function does.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    saved_image_paths.append(
+                        "Your request activated the API's safety filters \
+                        and could not be processed. Please modify the prompt and try again.")
+                if artifact.type == generation.ARTIFACT_IMAGE:
+                    saved_image_paths.append(self.save_image(artifact))
+        return saved_image_paths
+    
+    def get_existing_engines(self):
+        url = "https://api.stability.ai/v1/engines/list"
+        if self.api_key is None: raise Exception("Missing Stability API key.")
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:45]: Please indent this properly.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        response = requests.get(url, headers={
+            "Authorization": f"Bearer {self.api_key}"
+        })
+
+        if response.status_code != 200:
+            raise Exception("Non-200 response: " + str(response.text))
+
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:53]: Consider using try-except block for this api request and raise an exception if status code is not 200.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        
+        answers = self.stability_api.generate(
+            prompt=prompt,
+            seed=seed,
+            height=height,
+            width=width,
+            steps=steps,
+            sampler=generation.SAMPLER_K_DPMPP_2M
+        )
+
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:76]: Similarly, use try-except, and raise an Exception stating `Cannot generate Image`.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+from mindsdb.utilities.log import get_log
+
+
+logger = get_log("integrations.stabilityai_handler")
+
+class StabilityAIHandler(BaseMLEngine):
+    name = "stabilityai"
+
+    @staticmethod
+    def create_validation(target, args=None, **kwargs):
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:17]: Mention in comments what this function does.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            raise Exception('local_directory_path has to be specified')
+        
+        client = StabilityAPIClient(args["api_key"], args["local_directory_path"])
+        
+        if 'stability_engine_id' in args:
+            if not client._is_valid_engine(args["stability_engine_id"]):
+                raise Exception("Unknown engine. The available engines are - " + list(client.available_engines.keys()))
+
+    def create(self, target: str, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+        if 'using' not in args:
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:36]: What is the purpose of this function? @lnxpy

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+        input_column = args['input_column']
+        
+        if input_column not in df.columns:
+            raise RuntimeError(f'Column "{input_column}" not found in input data')
+        
+        local_directory_path = args["local_directory_path"]
+        stability_engine_id = args.get('stability_engine_id', "stable-diffusion-xl-1024-v1-0")
+        
+        client = StabilityAPIClient(api_key=api_key, dir_to_save=local_directory_path, engine=stability_engine_id)
```
Comment (tcchawla) [inline mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:55]: Consider using try-except block for the API call.

**Review by tcchawla — CHANGES_REQUESTED** [review]
Hey @lnxpy, great work, please make these minor changes.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py (hunk 9 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:36]: @tcchawla This function is called when `CREATE MODEL` query is executed for the ML handler.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 2 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:24]: What's the benefit of raising an exception after catching it?

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 1 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:23]: Done

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 3 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:47]: Done

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 4 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:61]: Done

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 5 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py]: Done

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 6 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:87]: What's the benefit of raising an exception after catching it?  The exception will anyway be shown on the frontend.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 7 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py]: What's the benefit of raising an exception after catching it?  The exception will anyway be shown on the frontend.

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py (hunk 8 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:18]: Done

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py (hunk 10 above)
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:56]: What's the benefit of raising an exception after catching it?  The exception will anyway be shown on the frontend.

**Comment by Better-Boy** [comment]
> I'm not sure if a `setup.py` file would come in handy for a handler at this point. All you need to have is a `requirements.txt` that contains your dependencies. This file might cause some misunderstandings. Some `setup()` parameters are valued wrongly such as `name`. What do you think @ZoranPandovski?!

Please let me know the corrections to be made as this is my first ML handler and I'm facing a problem - https://mindsdbcommunity.slack.com/archives/C04EWFQEFEV/p1698317162200199

**Comment by Better-Boy** [comment]
The issue I'm having is as follows:

Operating System : MacOS

On trying to create a model, the model creation gets stuck at the generating phase. It never goes to a completion state. And also I get `null` in the model status bar as follows:

Screenshot 2023-10-26 at 4 15 32 PM

Creation of the stabilityai engine runs fine without any issues. Because of the above issue, the model never gets created.

**Comment by Better-Boy** [comment]
Currently there is no way to display images as response from the handler. Hence, the work around for now after discussion with @StpMax is - the image response will be stored in the local file directory and the path to the locally saved image will be displayed on the UI. Once `column_types` feature gets implemented, we can display images on the frontend. The conversion is captured here - https://mindsdbcommunity.slack.com/archives/C04EWFQEFEV/p1697112709886209

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
I was able to get this to work on my Linux (Ubuntu) machine,
image

The image gets generated as well,


@Better-Boy There is one change I think you need to make; let's add a task parameter similar to the HuggingFace (and HuggingFace) API handlers. I say this because I believe Stability AI supports a few other tasks, not just text to image.

Furthermore, I think how useful this integration is will depend on the front-end team implementing the feature to display images in the UI and of course, the ability to download them. I will open an issue for this.

@paxcema Any thoughts?

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py (hunk 6 above)
Comment (paxcema) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai.py:87]: This is fine for now I'd say, but up to you in the end @Better-Boy (same with all other comments that ask for this).

File: mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py (hunk 9 above)
Comment (paxcema) [reply mindsdb/integrations/handlers/stabilityai_handler/stabilityai_handler.py:37]: @tcchawla hopefully that would be clear from the fact this is a `BaseMLEngine`?

**Comment by Better-Boy** [comment]
> @Better-Boy There is one change I think you need to make; let's add a task parameter similar to the HuggingFace (and HuggingFace) API handlers. I say this because I believe Stability AI supports a few other tasks, not just text to image.

@MinuraPunchihewa I have added the support for `task` and integrated 3 other APIs of the stability. Thanks for this suggestion.

File: mindsdb/integrations/handlers/stabilityai_handler/README.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+Stability AI ML handler for MindsDB provides interfaces to connect with Stability AI ML via APIs and pull Stability AI ML Capabilites into MindsDB.
+
+## Stability AI
+
+Stability AI world’s leading open source generative AI company. Learn more about stability AI[here](https://stability.ai/about)
+
+## Implemented Features
+
+- [x] Stability AI ML Handler
+  - [x] Support Generate Image
```
[the remark this answers is not in the document]
Comment (Better-Boy) [reply mindsdb/integrations/handlers/stabilityai_handler/README.md]: @paxcema done, updated the readme.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
@Better-Boy This looks great now! As a final step, can you please provide some unit tests for this integration? You can put it in here,
https://github.com/mindsdb/mindsdb/tree/staging/tests/unit/ml_handlers

If you need help running these tests (since you've had trouble on your Mac), let me know. I will run them on my machine.
Please note that we have recently introduced a couple base test classes, so if you can use them when writing your unit tests that would be great,
https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/base_ml_test.py

You can use this as a reference if you like,
https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_anthropic.py

**Comment by Better-Boy** [comment]
> @Better-Boy This looks great now! As a final step, can you please provide some unit tests for this integration? You can put it in here, https://github.com/mindsdb/mindsdb/tree/staging/tests/unit/ml_handlers
> 
> If you need help running these tests (since you've had trouble on your Mac), let me know. I will run them on my machine. Please note that we have recently introduced a couple base test classes, so if you can use them when writing your unit tests that would be great, https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/base_ml_test.py
> 
> You can use this as a reference if you like, https://github.com/mindsdb/mindsdb/blob/staging/tests/unit/ml_handlers/test_anthropic.py

@MinuraPunchihewa I have added sample tests. Please run them on your machine and let me know if changes are needed.

**Review by MinuraPunchihewa — APPROVED** [review]

**Pull request merged**